include/bitcoin/system/wallet/witness_address.hpp

```diff
@@ -0,0 +1,169 @@
+class BC_API witness_address
+  : public payment_address
+{
+public:
+    enum class encoding: uint8_t
+    {
+        // BIP 142 p2wpkh and p2wsh address versions.
+        // https://github.com/bitcoin/bips/blob/master/bip-0142.mediawiki#specification
+        mainnet_base58_p2wpkh = 0x06,
+        testnet_base58_p2wpkh = 0x03,
+        mainnet_base58_p2wsh = 0x0a,
+        testnet_base58_p2wsh = 0x28,
+
+        // These encoding values are not from a specification.
+        mainnet_p2wpkh,
+        mainnet_p2wsh,
+        testnet_p2wpkh,
+        testnet_p2wsh,
+
+        unknown
+    };
+
+    static const std::string mainnet_prefix;
+    static const std::string testnet_prefix;
+
+    typedef std::vector<witness_address> list;
+    typedef std::shared_ptr<witness_address> ptr;
+
+    /// Create base58 witness programs using provided information.
+    static void to_witness(witness_p2wpkh& out, encoding out_type,
+        uint8_t witness_version, short_hash hash);
+    static void to_witness(witness_p2wsh& out, encoding out_type,
+        uint8_t witness_version, hash_digest hash);
+
+    /// Constructors.
+    witness_address();
+    witness_address(witness_address&& other);
+    witness_address(const witness_address& other);
+    witness_address(const witness_p2wpkh& decoded,
+        encoding out_type=encoding::unknown);
+    witness_address(const witness_p2wsh& decoded,
+        encoding out_type=encoding::unknown);
+    witness_address(const data_chunk& data,
+        encoding out_type=encoding::unknown);
+    witness_address(const std::string& address,
+        encoding out_type=encoding::unknown);
+    witness_address(short_hash&& hash, encoding out_type=encoding::unknown,
+        uint8_t witness_version=0);
+    witness_address(const short_hash& hash, encoding out_type=encoding::unknown,
+        uint8_t witness_version=0);
+    witness_address(hash_digest&& hash, encoding out_type=encoding::unknown,
+        uint8_t witness_version=0);
+    witness_address(const hash_digest& hash,
+        encoding out_type=encoding::unknown,
+        uint8_t witness_version=0);
+    witness_address(const chain::script& script,
+        encoding out_type=encoding::unknown);
+    witness_address(const ec_private& secret,
+        encoding out_type=encoding::unknown);
+    witness_address(const ec_public& point,
+        encoding out_type=encoding::unknown);
+
+    /// Operators.
+    bool operator<(const witness_address& other) const;
+    bool operator==(const witness_address& other) const;
+    bool operator!=(const witness_address& other) const;
+    witness_address& operator=(const witness_address& other);
+    friend std::istream& operator>>(std::istream& in, witness_address& to);
+    friend std::ostream& operator<<(std::ostream& out,
+        const witness_address& of);
+
+    /// Cast operators.
+    operator bool() const;
+
+    /// Serializer.
+    std::string bech32(const std::string& prefix=mainnet_prefix) const;
```

Comment:
This method should not be provided as it allows one to change the prefix for an already constructed address. The address object should contain all information upon construct. This also implies there is no use for the `unknown` version value.